### ISSUE TITLE
[Reproducibilty] image creation time, EPOCH, and USE_CURRENT_TIMESTAMP on Jib's reproducible, content-addressable, and declarative image building

### ISSUE THREAD

**Issue by wu105** [GoogleContainerTools/jib#2534]
**Description of the issue**:

this is a follow up on #1079 -- when an image was built with pre-configured timestamp, would like to find out actually when it was built.  This information may be used to diagnose build environment issues and cleanup images based on build time.

**Comment by chanseokoh** [comment]
Hi @wu105,

#1079 is about _file_ timestamps. I think you're talking about image creation time. You can [use `USE_CURRENT_TIMESTAMP`](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#why-is-my-image-created-48-years-ago) to set the current build time as your image creation time.

**Comment by wu105** [comment]
We need to know the build time when the `USE_CURRENT_TIMESTAMP` option is not used.
This can be a Docker issue because Docker has the information internally.
However, when the image uses the build time as the created time, the build time is readily available.

When we start to set the Create time to something else for reproducibility, the build time becomes hidden.  A provision need to be arranged to recover the build time.

**Comment by chanseokoh** [comment]
> This can be a Docker issue because Docker has the information internally.

@wu105 I'm still struggle to see what you are looking for. I don't know what you mean by "Docker" above, but when it comes to building a container image, there's no "hidden" or "internal" attributes or properties associated with an image that is less easy to read or process for any user or tool. An image is merely a collection of tar archives plus a JSON file (a.k.a. "image configuration") that contains image metadata including the image creation time. And this JSON image metadata has a well-defined spec, for example, [OCI image configuration](https://github.com/opencontainers/image-spec/blob/master/config.md#properties). When you build or process an image, you are always creating or reading the JSON image configuration, and there's nothing "hidden."

But do note that you can associate arbitrary values to an image by utilizing "labels." (Check `Labels:` in [this example](https://github.com/opencontainers/image-spec/blob/master/config.md#example).) However, if you are going to store changing timestamps in the image configuration, understand that it will affect the reproducibility just as `USE_CURRENT_TIMESTAMP` will do. So I think the only and the best option for you is to use `USE_CURRENT_TIMESTAMP`.

**Comment by wu105** [comment]
@chanseokoh When the `USE_CURRENT_TIMESTAMP` option is used, `docker images` would tell me when an image is built via the attribute `CREATED`.   How do I know when an image is built when the  `USE_CURRENT_TIMESTAMP` option is *not* used?  Certainly not by adding a label that would defeat the purpose of not using the `USE_CURRENT_TIMESTAMP` option in the first place.

Docker manages information not considered part of the image, e.g., repositories and tags.  When we change the meaning of the `CREATED`  time to make the image reproducible, we should redesign `docker images`, `docker inspect`, `docker history` and other commands to show the build time.  Part of the problem with reproducing an image is that not enough information is recorded/displayed and we may need to go back to the drawing board for reproducing images.

**Comment by chanseokoh** [comment]
> How do I know when an image is built when the `USE_CURRENT_TIMESTAMP` option is _not_ used? Certainly not by adding a label that would defeat the purpose of not using the `USE_CURRENT_TIMESTAMP` option in the first place.

There is no way you can attach some information (I believe real-world build time in your case) to an image, unless you really attach that information to an image (whether for `CREATED` or labels), obviously. And when you attach a piece of information, that information becomes part of the image that affects the signature (SHA256 checksum hash) of the image (i.e., reproducibility). There is no way you can natively inject or associate some information that does not affect the image checksum. If you do want to attach to the image the information of when actually the image was built, I think the best option is `USE_CURRENT_TIMESTAMP` as I said.

Otherwise, the only option I can think of is to set up your own database outside of the container ecosystem where you manually map image digests and their actual real-world build timestamps and manage the database yourself.

> Docker manages information not considered part of the image, e.g., repositories and tags.

I am still not sure what you mean by "Docker" here. Are you talking about the Docker engine (also called Docker daemon) that usually runs locally on a developer's machine? Note that a Docker engine is not a container registry. By "repositories and tags", I believe you meant an [image manifest](https://github.com/opencontainers/image-spec/blob/master/manifest.md), but it does have any place to store arbitrary information. Even if it had a place to associate some information, it would affect the SHA256 checksum of the manifest itself just like the image configuration. I don't know if there exist some registry implementations that may save additional metadata of stored images allow the user to query that kind of information via their own proprietary API. However, even in that case, such features would not be part of the Docker/container specifications.

**Comment by chanseokoh** [comment]
Correction: it does not* have any place to store arbitrary information either.

**Comment by wu105** [comment]
Well, if the CREATED field can re-purposed (in effect it is as if removed), we might as well create something to store the build time in places similar to tags.  On the original build server, there are more information the docker client can access then it displays.  Similarly, the registry stores some useful information that would be useful if can be retrieved by a docker client, such as push time, pull counts.  More information that is only available via API could also be added to the docker client.

**Comment by chanseokoh** [comment]
Sorry, I am still very confused about what exactly you expect from Jib.

> we might as well create something to store the build time in places similar to tags.

Who exactly are "we" in this context? Your organization/dev teams? Jib? Or the Docker/container specification?

And are you making a new proposal to change the Docker/container or Docker Registry API specifications for the entire container ecosystem in some way or another? Or are you suggesting a potential workaround that can work in your situation (e.g., just have tags embed a timestamp)?

> the original build server

What exactly is a build server? I have difficult making connection to Jib in this context.

> there are more information the docker client can access then it displays. Similarly, the registry stores some useful information that would be useful if can be retrieved by a docker client, such as push time, pull counts. More information that is only available via API could also be added to the docker client.

As you keep talking about "Docker" and "docker client", it makes me think maybe if it isn't the case that you have mistaken a different repo for what you are asking? (Sorry, if not.) This repo is for the Jib tool set, and not a repo for requesting changes to the Docker/container specifications. As I said earlier, a registry may store some additional information that can be useful to you (like push time, pull counts, etc). However, it's not something Jib can/will make use of it, as it is just not part of the Docker/container specifications. Whatever you are suggesting here, I think it is just out of the scope of Jib.

Again, I am still very confused about what are you are asking or what you are trying to achieve in the end in Jib. However, whatever it is, I think there is nothing Jib can really change here.

**Comment by wu105** [comment]
It is a good question as to who is responsible for this issue.
It should be the one who created the  `USE_CURRENT_TIMESTAMP` option that essentially deleted the CREATED field from images.  I created this issue as a follow up to  #1079

**Comment by chanseokoh** [comment]
@wu105

> It should be the one who created the USE_CURRENT_TIMESTAMP option that essentially deleted the CREATED field from images.

Probably not. I don't think that your argument above applies here because
- Jib has in no way "deleted" the CREATED field, whether effectively or essentially. The CREATED field is the inherent config property of a container image; it is strictly part of the image spec. And Jib gives the user the total and absolute control, liberty, and authority over what value they want to store in the CREATED field. Jib imposes zero restrictions on how the CREATED field can be set or used, and Jib does not alter the value set by the user in any possible way. 
- Jib does not dictate, whether implied or not, how the CREATED field should be interpreted or try to grant any Jib-specific semantics to the values in the CREATED field set by the user. It is up to the user to decided how they want to interpret the value of the CREATED field for their purposes.
- `USE_CURRENT_TIMESTAMP` is merely an optional convenience shortcut; even if it didn't exist, anyone could still set any value to the CREATED field, whether it is the current system time or not. The shortcut just makes people's lives much easier by having Jib retrieve the current time on behalf of them; without the shortcut, the user would need to manually retrieve the current time from the system and pass it to Jib themselves.

> I created this issue as a follow up to #1079

As I said in https://github.com/GoogleContainerTools/jib/issues/2534#issuecomment-646631753, #1079 is about last modification time of the files containerized in an image. That is, #1079 is independent of the CREATED field you've been talking about. So I think you are actually confused, and this is probably the reason I really have hard time understanding what exactly you want from Jib.

> It is a good question as to who is responsible for this issue.

I'm still having trouble understanding what the "issue" is. But so far, my understanding is that
- You want to store whatever timestamp you want for your purpose (strictly speaking, the current system time when Jib runs, i.e., the "build time" as you refer to) in the built image.
- However, you do not want to store the timestamp in the CREATED field, because you will store a different timestamp (whatever it is) in the CREATED field for a different purpose. That is, your requirement is that you want to store _two_ different timestamps in the image. Obviously, you can store the first value in CREATED. Now you are looking for a different field than CREATED to store the second value. And you said image labels are not the field you want to use to store the second value.
- When storing this second value in some field, you wish this field would not impact the reproducibility of the image. However, as I have stressed a few times, this is not achievable.

**Comment by wu105** [comment]
@chanseokoh I think we failed to connect.  To me, "CREATED" has always been the build time of the image.  Once arbitrary value is placed there, the "build time" is lost.  My first priority is to recover the "built time" somehow, trying not to address whether such a practice is proper although this is a legitimate question.

One thing that might help is to go back to the definition of the image's configuration properties.    Is https://github.com/moby/moby/blob/master/image/spec/v1.md an acceptable copy?  Here is a quote from it:
```
created string
  ISO-8601 formatted combined date and time at which the image was created.
```

By the way, I am talking mostly images built by *others*, and I have no control over what value is there.  I have seen a lot of images "CREATED" 1970-01-01T00:00:00Z, which screams "I am not the create date purported to be", thus the remark that this field is in effect being deleted in these images.  I think we got a problem here in the real world regardless whether this is the intention of the designer of the `USE_CURRENT_TIMESTAMP` option.

**Comment by chanseokoh** [comment]
@wu105 I think now I understand part of your intention to some degree (still not 100%). I thought you were a developer and Jib user, but now I get that the context is that you are basically in an org that primarily manages container images built by others and not happy seeing images built by some tools like Jib (as they often have `1970-01-01T00:00:00Z` in CREATED.)

> To me, "CREATED" has always been the build time of the image. Once arbitrary value is placed there, the "build time" is lost. ... trying not to address whether such a practice is proper although this is a legitimate question.

This kind of topic has been already fiercely debated by people for a very long time in many contexts. As an example, [reproducible-build.org](https://reproducible-builds.org) (by which we've been inspired as well) has the section with the title [_“Lying about the time” / “violates language spec”_](https://reproducible-builds.org/docs/source-date-epoch/). You can read the entire section which is fairly long, but to quote some part, it says

> **_“Lying about the time” / “violates language spec”_**

> ... In most cases, this argument places too much weight on the absoluteness of time. Regardless of what any specification says, the user can set their own system clock and achieve an effect similar to SOURCE_DATE_EPOCH. ... It is not up to the tool to judge whether the user is lying with their system clock, and likewise, it is not up to the tool to judge whether SOURCE_DATE_EPOCH is a “lie” or not.

> For all intents and purposes, if the user has set SOURCE_DATE_EPOCH then they are taking a position that “this is the current time; please use this instead of whatever clock you normally use”. Yes, the project developer wrote “get_current_time()” but I as the user, by setting SOURCE_DATE_EPOCH, am choosing to override this with my own idea of what time it is. Please execute the build as if the current time was SOURCE_DATE_EPOCH. ...

> In conclusion, ... to judge that this is a ‘‘lie’’ is to disrespect the user’s wishes. Furthermore, choosing to support this is unlikely to ‘‘actually’’ violate any specifications, since they generally don’t define “current”. ...

> Many tools allow the user to override the “current” date - e.g. -D__TIME__=xxx, \\year=yyy, etc. In these cases, it makes even less sense to ignore SOURCE_DATE_EPOCH for data integrity reasons - you ‘‘already’’ have a mechanism where the user can “lie” or “break the spec”; ...

You can still disagree on the arguments like the one above. If so, I'd say we can agree to disagree with each other.

Another argument is the usual file modification time on any OS. Someone can create a `foo.txt` and then manipulate its last modification time to whatever value they wish. In a similar vein, Jib as a tool will respect the user's wish–they can create an image and then manipulate its CREATED.

So, now it looks like your intention and goal as I understood here is to try to influence the design direction of Jib so that Jib will always use the current system clock to set the CREATED value, or at least try to change the default behavior of Jib so that it doesn't set `1970-01-01T00:00:00Z` by default. However, I don't think we will change our current design on this part. You may not agree on our decision, but I am afraid we will continue to give the user the total control and authority over whatever value they wish to set for CREATED. I think we should agree to disagree.

In conclusion, now I see there is no actionable item for us regarding this "issue", which we don't think is an actual issue. But I appreciate your feedback. We made a long detour to finally connect to each other (although I still don't think we're 100% connected), but I think our discussions with a long detour have been constructive.

**Comment by wu105** [comment]
In terms of opinions, I am not sure when would be a good use case for reproducing an image.  All I can think of is to calibrate a build environment to be *able* to reproduce exactly the same image then *move on* to use the environment for building a *improved* image.  In this situation the reproduced image should *never* need to leave the build environment, because we always have the original image, otherwise how can we be certain it *is a carbon copy*?

I am certain the designer of the 'USE_CURRENT_TIMESTAMP' option is for it to be set to the build time when an image is initially built, and set to a different time *only when reproducing*.  This would still be in agreement with the image configuration specification.

However, there is a prevalence of using `SOURCE_DATE_EPOCH` for CREATED, all on images that were *built for the first time*.   Using EPOCH don't gain anything over using build time.  We probably really need to get the  image's configuration specifications amended to make images reproducible while we have a standard on CREATED.

**Comment by chanseokoh** [comment]
@wu105

> In terms of opinions, I am not sure when would be a good use case for reproducing an image. All I can think of is to calibrate a build environment to be _able_ to reproduce exactly the same image then _move on_ to use the environment for building a _improved_ image.

That's a fair opinion, I admit. However, in the context of container images, there are many more reasons for the importance and necessity of reproducilbity than you are seeing.

One notable example is the [content-addressability](https://docs.docker.com/registry/spec/manifest-v2-2/) of container images. Generally speaking, content-addressability means that, if you have a content of something, you and everyone else immediately and automatically know or are able to get the address of the location where the content lives. Likewise, in the opposite direction, if you have obtained an address, not only can you locate the content by the address but also can be assured that the content you will find at the address will always be the one you know or expect even without actually fetching the content. For example, even if the CREATED value is the only field you changed, you will be assigned a new foreign address for the image, and that matters a lot in the container image ecosystem.

Another increasingly popular image management practice that we hear a lot from actual Jib users is a modern GitOps practice. For example, when composing an image, they set to CREATED the last modification timestamp of the git commit of an image source repo from which the image would be built. For them, where and when in the real world a physical form of an image corresponds to a certain commit can be built and stored where and how many times is unimportant; rather, what is more important is the reproducibilty in the context of the GitOps paradigm, where the source repo is the ultimate source of truth.

I can go over and present more examples, but I don't think I really need to go further. Just understand that there are many more reasons for the strong desire for reproducibility especially in container images, and that different people have different perspectives and opinions on what CREATED means for them. We (and Jib) do not go into the business of judging which opinion is right or wrong; we respect all the perspectives (including yours) and give the total liberty on what they want to do with CREATED. It is then up to the user to decide if they will put a real-world build time or a fixed timestamp.

> I am certain the designer of the 'USE_CURRENT_TIMESTAMP' option is for it to be set to the build time when an image is initially built, and set to a different time _only when reproducing_. This would still be in agreement with the image configuration specification.

I am sure by now you have already guessed correctly, but just to be clear: the answer is no. :smile: We started with setting the epoch and then designed and introduced the `USE_CURRENT_TIMESTAMP` later, and when we did it, we had a clear plan and intention. I don't think I have to repeat myself about what our intention was.

> We probably really need to get the image's configuration specifications amended to make images reproducible while we have a standard on CREATED.

I had the same question and sentiment for this unfortunate situation in the past. I still do. Couldn't agree more. However, I see that some inherent choices made by the container image designers like the content-addressability make this difficult, unfortunately.

**Comment by wu105** [comment]
@chanseokoh Thanks for the information.   You do not need to repeat past arguments, some links would be sufficient.

I am still not get it that the "build time" has to be made unavailable to achieve the goals.  For instance, if the idea is that `the image’s configuration can be hashed to generate an ID for the image`, how can one deduce that the creation time can no longer be part of the image's configuration unless it is EPOCH? 

When the CREATED is rendered *meaningless*, e.g., set to EPOCH, provisions should and could be made to have the "build time" available elsewhere.  It would be a problem if there  is no standard or no support/enforcement of the standard from the tools.

**Comment by chanseokoh** [comment]
> I am still not get it that the "build time" has to be made unavailable to achieve the goals.

I think it isn't that "build time" _has to_ be made unavailable. I'd say it is up to the user depending on their goals. Some decide to forgo the reproducibllity and just use a real-world "build time" in CREATED for their goals. Others use their own meaning of creation time for their goals, preserving reproducibility. We (and Jib) don't define the goals or judge if their decisions are right.

> When the CREATED is rendered meaningless, e.g., set to EPOCH, provisions should and could be made to have the "build time" available elsewhere.

Sorry to repeat myself, but as I said right above, I am mostly empathetic on this unfortunate situation, and as much as you do, I really wish there existed an ideal solution.

> It would be a problem if there is no standard or no support/enforcement of the standard from the tools.

Again, sorry to repeat, but I believe this is where I said we disagree. Our tool will not force a particular standard on CREATED but rather let the user, tools, apps, and systems define their semantics and interpretations of what CREATED means for them.

I think our discussions are now circling around, and seems like I have to repeat myself multiple times. What's obvious to me though, is that whatever the ideal world you'd like to see, this is out of scope of Jib unfortunately. I will close this issue as "out of scope" and "working as intended." Personally, I think it would be a more effective use of your time if you start this kind of discussions with the Docker/OCI folks publishing their specs rather than complaining the lack of some kind of ecosystem support here in this repo. (FYI, just in case you start a new discussion with them, OCI and Docker specs are mostly identical but they do have divergence). But as always, the Jib dev team welcomes any constructive feedback and discussions.

**Comment by wu105** [comment]
@chanseokoh Would you please provide some links to the past discussion threads?   Also links to recommendations on choosing CREATED time?  I would like to read these before discussing with our image owners about selecting the CREATED time.

I have navigated around in the oci site https://opencontainers.org/, checked all issues at https://github.com/opencontainers/image-spec/issues, joined the http://chat.opencontainers.org/ and searched, got 18 hits searching `USE_CURRENT_TIMESTAMP` in https://github.com/GoogleContainerTools/jib/issues, plus all the googling upto the current point, but did not find any.   

I have noticed that you have closed the issue and directed me elsewhere.  However, the remaining seems relevant for JIB and this is an issue on the JIB issue list after all.  If it is better to create different issues or they are already addressed, please let me know.

Judging from the issues turned up googling, users of jib seems surprised by images older than they are, one issue even enumerated from 48 years old to 55 years old,  instead of having any purpose in setting EPOCH as CREATED date.  In a minimum I think, the default should be changed to some meaningful build time, e.g., git commit time, until the users purposely opt in for a different time.   jib can print encouragements for users to do so.

Readiing the following at https://github.com/GoogleContainerTools/jib
```
Goals
Reproducible - Rebuilding your container image with the same contents always generates the same image. Never trigger an unnecessary update again.
```
I wondering would it be better to skip a new build  when it would not produce a different image?
If it is impossible to know without actually rebuild the image, i.e., the image is not at content-addressable yet, the new build could use the CREATED time *from the existing image*.
If the newly build is different, the newly build can update its CREATED before publishing, otherwise it can be discarded.
This might even be faster and more efficient than building and pushing an identical image unconditionally.
Is this too much a price for making the CREATED time more meaningful?
I think we can always try to do better than "pulling a CREATED time from the thin air", even if we have to do it case by case.
We may have all the bases covered sooner than expected if we continue this way.

**Comment by chanseokoh** [comment]
@wu105

> I have noticed that you have closed the issue and directed me elsewhere.

I can always reopen the issue. Just to be clear, I genuinely believed and thought that what you have been asking for is out of scope of Jib, and Jib is working as intended, and the OCI/Docker folks are the right people for this kind of discussions. (I still believe so, to be honest.)

> I wondering would it be better to skip a new build when it would not produce a different image?
If it is impossible to know without actually rebuild the image, i.e., the image is not at content-addressable yet, the new build could use the CREATED time from the existing image.
If the newly build is different, the newly build can update its CREATED before publishing, otherwise it can be discarded.
This might even be faster and more efficient than building and pushing an identical image unconditionally.

Let me first take care of these questions, as I think doing so will help you shape a better understanding of the reproducibility we have been talking about. Reproducibility is not just about skipping a second rebuild or pushing if you already completed a build before. To me, most of your questions sound out of context and don't seem to apply well into Jib's way of thinking; I'll explain why shortly.

Throughout our discussions, I had noticed that you have established a rigid mental model where you tend to develop your thoughts in terms of certain notions like "initially built image", "first image", "existing image", "reproducing an original image", "rebuilding", etc. The reason you were able to conceive these questions above at the first place is because you are perceiving the world around you through the lens of the mind where things are mutable, imperative, and physical. That is, in your world, things are constantly mutated, the order of events are critical, the act of actually initiating a build matters, the act of actually saving image content bytes into a physical storage and holding them matters, etc.

For those who truly understand and embrace reproducibility–basically Jib's way of image building, their mental model is quite different: immutable, declarative, and conceptual. Let me give you a concrete example:

Suppose you have just set up a fresh new repo for a Java hello-world app. Even at this very early stage, even before thinking anything about containerizing the app, in the world perceived through Jib's lens is an omni-present, "conceptual image" tied with the app that has not even ever been compiled once. It is then only a matter of incarnation–a matter of when to materialize the image at a future point when deemed necessary. For example, let's say Alice checks out the repo into her MacBook, and Bob checks out the same repo on his Windows desktop. Alice runs Jib and builds the image into Docker Hub. Bob runs Jib several months later and builds into Quay.io. With the strong reproducibility that Jib provides, third-party container tools capable of comparing images byte-to-byte will report that these two images in Docker Hub and Quay.io uploaded at different times are absolutely identical (when `USE_CURRENT_TIMESTAMP` is not used, of course). It doesn't matter to Alice or Bob if both uploaded images are deleted from the registries, if there are no registries in the world holding the image, where some materialized images are stored, or if they run Jib multiple times in other centuries. All the images built by Jib are guaranteed to be the same image byte-to-byte. In this perspective, notions like "use the CREATED time from the existing image", "update its CREATED", "before publishing", or "can be discarded" don't make much sense. Whenever and as soon as Alice or Bob creates a new Git/SVN/Mercury/... commit to the source repo, the corresponding conceptual image is already "created and existing" in this world. Therefore, in this sense, it would make sense to regard the Git/SVN/Mercury/... commit timestamp as the creation time of conceptual images.

> Is this too much a price for making the CREATED time more meaningful?

I believe I explained this before. For some, it's a prohibitive price they can never take. For others, not. It depends. BTW, "meaningful" is subjective, if you think about it. For you, having a real-world time is more meaningful. For others, using a fixed timestamp is more meaningful. Jib cannot make a judgement and has been leaving the decision to them.

> Would you please provide some links to the past discussion threads?

My guess here is that you are referring to discussions around `USE_CURRENT_TIMESTAMP`? If so, unfortunately, there are no public discussion threads that I see; this option was designed and implemented into Jib at the very early stage of Jib development long before Jib went GA. However, [this video talk](https://youtu.be/H6gR_Cv4yWI?t=1074) (which is really old and largely out of date as of now) briefly talks about wiping out some metadata including timestamp. The talk doesn't bother to mention `USE_CURRENT_TIMESTAMP`, but at least you should be able to confirm our intention we had from the beginning regarding the option. As I said, we started our project with strong emphasis on reproducibility in mind, but at the same time, the option was a necessity as well to support use-cases like yours.

> Also links to recommendations on choosing CREATED time? I would like to read these before discussing with our image owners about selecting the CREATED time.

We do not have such an official link with specific recommendations, as we are not so heavily opinionated when it comes to choosing the CREATED time. As I said, we mostly leave the decision up to the user when they know what they are doing. However, the closest thing we have is the FAQ link I shared in my first comment. You'll see that the FAQ explains how to configure CREATED and also talks a bit about the impact on reproducibility, but it does not give any strong opinion on the matter. However, whenever we have opportunities, we generally encourage in an informal manner to maintain strong reproducibility by using a fixed timestamp. Using `USE_CURRENT_TIMESTAMP` makes it practically infeasible to verify whether two images built from the same source at different times and different places are identical (which often matters a lot to folks who take verifiabilty seriously for security or prod-dev parity reasons, for example). And considering the mental model of the conceptual image creation, IMO, it makes a lot of sense and brings many advantages to use a Git/SVN/Mercury/... commit timestamp as CREATED, although any fixed timestamp meaningful to you works too, for example, a timestamp of an important file to be put into the image. However, because using `USE_CURRENT_TIMESTAMP` is not expensive at all in terms of utilizing storage and network bandwidth in practice, we also have no problem suggesting the option for those who ask for it and don't care much about strong reproducibilty and an insignificant inefficiency in utilizing storage space.

That said, in your situation, in my opinion, because you seem to be so averse about putting any non-real-world time into CREATED, in that case, I should suggest to use `USE_CURRENT_TIMESTAMP`. However, in general we think it's better to use a fixed timestamp of your choice, which could be a commit timestamp of whatever VCS you use.

**Comment by wu105** [comment]
@chanseokoh Appreciate the explanation.  Is there a clearing house on this topic?  If so I would be happy to move the discussion there.

On OCI/Docker redefining image CREATED, before the specification is changed, I would prefer to stay true/close to it when possible.   Time stamp is one of few universal attribute of artifacts that may be best to be readily available.

You are right about my mind frame.  Indeed the problem only arises once we have an image built.  Your example is right to the point that independently built images can be easily determined to be the same.  However, that would require the parties to use the same CREATED time currently, and the JIB default EPOCH would make it likely to have this requirement meet.  However, using the git commit time as the default should have the same result statistically, and the git commit date would still be close to the spirit of image's CREATED.  If we have to lie about time, it is better to make good lies:-)

It is possible that one image builder have stripped the git metadata in the build environment, and setting CREATED to EPOCH would still work while setting CREATED to git commit date would not.  But do we really want to build images in parallel out of version control?  We have enough problems to build from even the same source control system.

In the byte compare example, if the two images differ, and we trust both and want to pick the newest, how do we decide when they have the same CREATED time?

From what I have learned so far, I can list the following CREATED candidates, in my preference order in light of supporting content-addressible images: source control commit time, max of source control commit time including upstream sources, CREATED of build and update to current time if resulting a new image, fixed time > EPOCH that increments for new images, fixed time >EPOCH, EPOCH.

I understand that you try not to make a recommendation.  However, having a clearing house of different selections for the CREATED time and the pros and cons is still valuable, at least I would not have to re-invent wheels here.  

By the way, selecting EPOCH as the default CREATED JIB is actually stronger than making a recommendation.

**Comment by loosebazooka** [comment]
@wu105 I think you make some valid points, but as @chanseokoh pointed out, this is a design decision that jib made *with a provision for user configurable alternative*. It appears now that this has become a philosophical issue which is not the purpose of our issue tracker. We've already committed to our approach for this particular configuration option.

Unfortunately it is out of scope of our team to address the build configuration of your build team. If you wish to enforce certain configurations on your builds at your company, there are certainly avenues to do so, but asking the jib team to change our defaults to match your expectations and business needs is probably not the way.

On top of that, some registries actually have an `uploaded` time feature which you can use to determine potentially when an image was built (as long as you're doing build + upload roughly around the same time). While this configuration is part of the registry system's metadata, and not built directly into the image manifest, it could be a solution to your problems.

**Comment by wu105** [comment]
@loosebazooka @chanseokoh There are some philosophical components in this, but first of all I am making two requests, and I am guilty of not have made them explicit:

- changing the default away from fixed CREATED time

- document the pros and cons of different CREATED time

On the philosophical side, one of the basics, at least of mine, is to minimize breaking changes when possible. JIB may have started for internal use, but images created by it are making way to unsuspecting registries.  The changes we have discussed can minimizing missing CREATED time *without* preventing you from doing what you set out to do nor asking you to make philosophical concessions.  I fact, I admire what you are exploring and appreciate the explanations.  

I understand that some registries, e.g., harbor, record pushing time along with other useful information, but I have not figured out whether there is an api for retrieving those bonus information.   The upload time won't help anyway for images not yet uploaded, re-tagged, or for the registries that I do not have access.  I have recently learned that one can listen to the docker events to recover historical information on the build machine, but that is way more complicated than checking the CREATED time.

**Comment by chanseokoh** [comment]
I know what I am going to tell you will disappoint you, but as I've iterated multiple times in this thread, I don't think Jib will change the current default for CREATED. We do understand your points and concerns that are valid and grounded, but when we set out this initiative a couple years ago, we made a conscious and reasoned design decision around this where some of the bases of the decision have been discussed in this thread. I hope you will understand.

Just an FYI, there are other image building tools that work in the same way as Jib, such as Bazel [`rules_docker`](https://github.com/bazelbuild/rules_docker). Also the public images from the popular [Distroless](https://github.com/GoogleContainerTools/distroless) project has the same default timestamp of epoch.

> document the pros and cons of different CREATED time

Great suggestion. I think this thread already serves as a nice reference (albeit too long and not organized), but I agree we can improve our documentation, perhaps by expanding [the FAQ entry](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#why-is-my-image-created-48-years-ago) for tracking purposes to update the doc. ls and summary on this front. We will leave this issue open for tracking purposes to update the doc.

---

For accurate documentation:

> Your example is right to the point that independently built images can be easily determined to be the same.

Determining whether multiple images are same is not just easy but instantaneous and immediate, _without comparison_, thanks to the [content-addressability](https://docs.docker.com/registry/spec/manifest-v2-2/) and strong reproducibility provided by Jib. That is,
   - same image address if any only if identical images
   - different image addresses if and only if different images

However, OTOH, it becomes very difficult and extremely cumbersome to verify in practice if two images are basically same when `USE_CURRENT_TIMESTAMP` is used which destroys reproducibility.

**Comment by chanseokoh** [comment]
For anyone coming to this thread and wanting to understand more about the image creation time and Jib's reproducible, declarative, and immutable image building, please carefully read through all the comments above as well as this Jib [FAQ](https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#why-is-my-image-created-48-years-ago).
